{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "95b6cb2b",
   "metadata": {},
   "outputs": [],
   "source": [
    "import unittest\n",
    "%run VCF_annotation_functions.ipynb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "f58727c9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# test response for all values returned by function get_info\n",
    "testresponse = \\\n",
    "[{'allele_string': 'A/G',\n",
    "  'assembly_name': 'GRCh37',\n",
    "  'colocated_variants': [{'allele_string': 'COSMIC_MUTATION',\n",
    "                          'end': 1158631,\n",
    "                          'id': 'COSV55420653',\n",
    "                          'phenotype_or_disease': 1,\n",
    "                          'seq_region_name': '1',\n",
    "                          'somatic': 1,\n",
    "                          'start': 1158631,\n",
    "                          'strand': 1,\n",
    "                          'var_synonyms': {'COSMIC': ['COSM3750257']}},\n",
    "                         {'allele_string': 'A/G/T',\n",
    "                          'end': 1158631,\n",
    "                          'frequencies': {'G': {'aa': 0.9789,\n",
    "                                                'afr': 0.9985,\n",
    "                                                'amr': 0.9294,\n",
    "                                                'sas': 0.956}},\n",
    "                          'id': 'rs6603781',\n",
    "                          'minor_allele': 'A',\n",
    "                          'minor_allele_freq': 0.0467,\n",
    "                          'seq_region_name': '1',\n",
    "                          'start': 1158631,\n",
    "                          'strand': 1}],\n",
    "  'end': 1158631,\n",
    "  'id': '1:g.1158631A>G',\n",
    "  'input': '1:g.1158631A>G',\n",
    "  'most_severe_consequence': 'synonymous_variant',\n",
    "  'seq_region_name': '1',\n",
    "  'start': 1158631,\n",
    "  'strand': 1,\n",
    "  'transcript_consequences': [{'amino_acids': 'D',\n",
    "                               'biotype': 'protein_coding',\n",
    "                               'cdna_end': 863,\n",
    "                               'cdna_start': 863,\n",
    "                               'cds_end': 570,\n",
    "                               'cds_start': 570,\n",
    "                               'codons': 'gaT/gaC',\n",
    "                               'consequence_terms': ['synonymous_variant'],\n",
    "                               'gene_id': 'ENSG00000078808',\n",
    "                               'gene_symbol': 'SDF4',\n",
    "                               'gene_symbol_source': 'HGNC',\n",
    "                               'hgnc_id': 24188,\n",
    "                               'impact': 'LOW',\n",
    "                               'protein_end': 190,\n",
    "                               'protein_start': 190,\n",
    "                               'strand': -1,\n",
    "                               'transcript_id': 'ENST00000263741',\n",
    "                               'variant_allele': 'G'},\n",
    "                              {'amino_acids': 'D',\n",
    "                               'biotype': 'protein_coding',\n",
    "                               'cdna_end': 822,\n",
    "                               'cdna_start': 822,\n",
    "                               'cds_end': 570,\n",
    "                               'cds_start': 570,\n",
    "                               'codons': 'gaT/gaC',\n",
    "                               'consequence_terms': ['synonymous_variant'],\n",
    "                               'gene_id': 'ENSG00000078808',\n",
    "                               'gene_symbol': 'SDF4',\n",
    "                               'gene_symbol_source': 'HGNC',\n",
    "                               'hgnc_id': 24188,\n",
    "                               'impact': 'LOW',\n",
    "                               'protein_end': 190,\n",
    "                               'protein_start': 190,\n",
    "                               'strand': -1,\n",
    "                               'transcript_id': 'ENST00000545427',\n",
    "                               'variant_allele': 'G'}]}]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "9996ff72",
   "metadata": {},
   "outputs": [],
   "source": [
    "# test response for all values returned unknown by function get_info\n",
    "testresponse2 = \\\n",
    "[{'allele_string': 'A/G',\n",
    "  'assembly_name': 'GRCh37',\n",
    "  'colocated_variants': [{'allele_string': 'COSMIC_MUTATION',\n",
    "                          'end': 1158631,\n",
    "                          'id': 'COSV55420653',\n",
    "                          'phenotype_or_disease': 1,\n",
    "                          'seq_region_name': '1',\n",
    "                          'somatic': 1,\n",
    "                          'start': 1158631,\n",
    "                          'strand': 1,\n",
    "                          'var_synonyms': {'COSMIC': ['COSM3750257']}},\n",
    "                         {'allele_string': 'A/G/T',\n",
    "                          'end': 1158631,\n",
    "                          'frequencies': {'G': {'aa': 0.9789,\n",
    "                                                'afr': 0.9985,\n",
    "                                                'amr': 0.9294,\n",
    "                                                'sas': 0.956}},\n",
    "                          'id': 'rs6603781',\n",
    "                          'minor_allele': 'A',\n",
    "                          #'minor_allele_freq': 0.0467,\n",
    "                          'seq_region_name': '1',\n",
    "                          'start': 1158631,\n",
    "                          'strand': 1}],\n",
    "  'end': 1158631,\n",
    "  'id': '1:g.1158631A>G',\n",
    "  'input': '1:g.1158631A>G',\n",
    "  #'most_severe_consequence': 'synonymous_variant',\n",
    "  'seq_region_name': '1',\n",
    "  'start': 1158631,\n",
    "  'strand': 1,\n",
    "  'transcript_consequences': [{'amino_acids': 'D',\n",
    "                               'biotype': 'protein_coding',\n",
    "                               'cdna_end': 863,\n",
    "                               'cdna_start': 863,\n",
    "                               'cds_end': 570,\n",
    "                               'cds_start': 570,\n",
    "                               'codons': 'gaT/gaC',\n",
    "                               'consequence_terms': ['synonymous_variant'],\n",
    "                               'gene_id': 'ENSG00000078808',\n",
    "                               #'gene_symbol': 'SDF4',\n",
    "                               'gene_symbol_source': 'HGNC',\n",
    "                               'hgnc_id': 24188,\n",
    "                               'impact': 'LOW',\n",
    "                               'protein_end': 190,\n",
    "                               'protein_start': 190,\n",
    "                               'strand': -1,\n",
    "                               'transcript_id': 'ENST00000263741',\n",
    "                               'variant_allele': 'G'},\n",
    "                              {'amino_acids': 'D',\n",
    "                               'biotype': 'protein_coding',\n",
    "                               'cdna_end': 822,\n",
    "                               'cdna_start': 822,\n",
    "                               'cds_end': 570,\n",
    "                               'cds_start': 570,\n",
    "                               'codons': 'gaT/gaC',\n",
    "                               'consequence_terms': ['synonymous_variant'],\n",
    "                               'gene_id': 'ENSG00000078808',\n",
    "                               'gene_symbol': 'SDF4',\n",
    "                               'gene_symbol_source': 'HGNC',\n",
    "                               'hgnc_id': 24188,\n",
    "                               'impact': 'LOW',\n",
    "                               'protein_end': 190,\n",
    "                               'protein_start': 190,\n",
    "                               'strand': -1,\n",
    "                               'transcript_id': 'ENST00000545427',\n",
    "                               'variant_allele': 'G'}]}]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "40ad3d79",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "test_get_hgvs_notation (__main__.testVCF_annotation) ... ok\n",
      "test_get_info (__main__.testVCF_annotation) ... ok\n",
      "test_get_variant_type (__main__.testVCF_annotation) ... ok\n",
      "\n",
      "----------------------------------------------------------------------\n",
      "Ran 3 tests in 0.021s\n",
      "\n",
      "OK\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<unittest.main.TestProgram at 0x21e2eff4880>"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# create test class\n",
    "\n",
    "class testVCF_annotation(unittest.TestCase):\n",
    "    \n",
    "    def test_get_variant_type(self):\n",
    "        self.assertEqual(get_variant_type('A','G'), 'substitution')\n",
    "        self.assertEqual(get_variant_type('A','AG'), 'insertion')\n",
    "        self.assertEqual(get_variant_type('ATTT','A'), 'deletion')\n",
    "        self.assertEqual(get_variant_type('AGCTTC','AGGTTA'), 'unknown')\n",
    "        \n",
    "    def test_get_hgvs_notation(self):\n",
    "        chromosome = '5' \n",
    "        position = 1234\n",
    "        placeholder = '1:g.25362501C>A'  # for unknown variant type\n",
    "        \n",
    "        self.assertEqual(get_hgvs_notation(chromosome, position,'A','G','substitution'), '5:g.1234A>G')\n",
    "        self.assertEqual(get_hgvs_notation(chromosome, position,'A','AG','insertion'), '5:g.1234_1235insG')\n",
    "        self.assertEqual(get_hgvs_notation(chromosome, position,'ATTT','A','deletion'), '5:g.1234_1237del')\n",
    "        self.assertEqual(get_hgvs_notation(chromosome, position,'AGCTTC','AGGTTA','unknown'), placeholder)\n",
    "        \n",
    "    def test_get_info(self):\n",
    "        self.assertEqual(get_info(testresponse), (['SDF4'], ['synonymous_variant'], [0.0467]))\n",
    "        self.assertEqual(get_info(testresponse2), (['unknown'], ['unknown'], ['unknown']))\n",
    "        \n",
    "    \n",
    "unittest.main(argv=[''], verbosity=2, exit=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "483645d7",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1b54a87f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "208bf43f",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
