{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "71ed1244",
   "metadata": {},
   "outputs": [],
   "source": [
    "# function to determine type of variant\n",
    "# inputs: reference base(s) and alternate base(s) from VCF\n",
    "# outputs: type of variant\n",
    "# currently, only identifies some variant types\n",
    "\n",
    "def get_variant_type(REF, ALT):\n",
    "    if (len(REF) == 1) and (len(ALT) == 1):\n",
    "        vartype = 'substitution'\n",
    "        \n",
    "    elif (len(REF) == 1) and (len(ALT) > 1):\n",
    "        vartype = 'insertion'\n",
    "            \n",
    "    elif len(REF) > len(ALT):\n",
    "        vartype = 'deletion'\n",
    "        \n",
    "    else:\n",
    "        # unknown variant type\n",
    "        vartype = 'unknown'\n",
    "        \n",
    "    return vartype"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "83fcc3d0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# function to put variant into hgvs notation\n",
    "# inputs:  chromosome, variant position, reference base(s), and alternate base(s) from VCF, \n",
    "# and variant type (from function get_variant_type)\n",
    "# outputs: hgvs notation\n",
    "# currently, only some variant types covered\n",
    "\n",
    "def get_hgvs_notation(chromosome, position, REF, ALT, varianttype):\n",
    "    # base\n",
    "    hgvs_part1 = chromosome + ':' + 'g.' + str(position)\n",
    "    \n",
    "    # variant specific\n",
    "    if varianttype == 'substitution':\n",
    "        hgvs_part2 = REF + '>' + ALT\n",
    "        \n",
    "    elif varianttype == 'insertion':\n",
    "        hgvs_part2 = '_' + str(position + 1) + 'ins' + ALT[1:]\n",
    "        \n",
    "    elif varianttype == 'deletion':\n",
    "        hgvs_part2 = '_' + str(position + len(REF) - 1) + 'del'\n",
    "        \n",
    "    else:\n",
    "        # unknown variant type\n",
    "        hgvs_part2 = ''\n",
    "        \n",
    "    if hgvs_part2 == '':\n",
    "        # put placeholder, ideally, all variant types put into proper hgvs notation\n",
    "        hgvs_notation = '1:g.25362501C>A'\n",
    "    else:\n",
    "        hgvs_notation = hgvs_part1 + hgvs_part2\n",
    "    \n",
    "    return hgvs_notation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ba7bce99",
   "metadata": {},
   "outputs": [],
   "source": [
    "# function to process Ensembl VEP hgvs API response from POST request\n",
    "# input: response from API as dict, processed by json.loads()\n",
    "# outputs: lists of gene symbol, most severe consequence, and minor allele frequency for each variant in input data\n",
    "\n",
    "def get_info(data):\n",
    "    gene = []\n",
    "    effect = []\n",
    "    minoralfreq = []\n",
    "\n",
    "    for item in range(len(data)):\n",
    "        # get gene, if available\n",
    "        if 'transcript_consequences' in data[item] \\\n",
    "        and 'gene_symbol' in data[item]['transcript_consequences'][0]:\n",
    "            gene.append(data[item]['transcript_consequences'][0]['gene_symbol'])\n",
    "        else:\n",
    "            gene.append('unknown')\n",
    "\n",
    "        # get effect, if available\n",
    "        if 'most_severe_consequence' in data[item]:\n",
    "            effect.append(data[item]['most_severe_consequence'])\n",
    "        else:\n",
    "            effect.append('unknown')\n",
    "\n",
    "        # get minor allele frequency, if available\n",
    "        if 'colocated_variants' in data[item] \\\n",
    "        and ('minor_allele_freq' in data[item]['colocated_variants'][0]):\n",
    "            minoralfreq.append(data[item]['colocated_variants'][0]['minor_allele_freq'])\n",
    "\n",
    "        elif 'colocated_variants' in data[item] \\\n",
    "        and (len(data[item]['colocated_variants']) > 1) \\\n",
    "        and ('minor_allele_freq' in data[item]['colocated_variants'][1]):\n",
    "            minoralfreq.append(data[item]['colocated_variants'][1]['minor_allele_freq'])\n",
    "        else:\n",
    "            minoralfreq.append('unknown')\n",
    "            \n",
    "    return gene, effect, minoralfreq"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
